lucet-concurrency-tests/src/killswitch.rs

```diff
@@ -0,0 +1,806 @@
+                    })
+                    .expect("can spawn a thread");
+
+                let terminator = in_guest.wait_and_then(move || {
+                    thread::spawn(move || {
+                        assert_eq!(kill_switch.terminate(), Ok(KillSuccess::Signalled));
+                    })
+                });
+
+                t.join().unwrap();
+                terminator.join().unwrap();
+
+                // Outstanding kill switches fail, because the kill state was reset.
+                assert_eq!(outstanding_killswitch.terminate(), Err(KillError::Invalid));
+            })
+        }
+
+        // Test that termination while entering a guest works without signalling the embedder.
+        //
+        // This corresponds to a race during the documentation's State A -> State B transition.
+        #[test]
+        fn terminate_entering_guest() {
+            let test_entering_guest_before_domain_change: fn(&Instance) -> SyncWaiter =
+                |inst: &Instance| -> SyncWaiter {
+                    inst.lock_testpoints
+                        .instance_entering_guest_before_domain_change
+                        .wait_at()
+                };
+            let test_entering_guest_after_domain_change: fn(&Instance) -> SyncWaiter =
+                |inst: &Instance| -> SyncWaiter {
+                    inst.lock_testpoints
+                        .instance_entering_guest_after_domain_change
+                        .wait_at()
+                };
+
+            for (i, racepoint_builder) in [
+                test_entering_guest_before_domain_change,
+                test_entering_guest_after_domain_change,
+            ]
+            .iter()
+            .enumerate()
+            {
+                println!("testing racepoint {}", i);
+                test_exclusive_instance_with_instrumented_guest_entry(|mut inst| {
+                    let kill_switch = inst.kill_switch();
+                    let racepoint = racepoint_builder(&inst);
+
+                    let guest = thread::Builder::new()
+                        .name("guest".to_owned())
+                        .spawn(move || match inst.run("run_guest", &[]) {
+                            Err(Error::RuntimeTerminated(TerminationDetails::Remote)) => {}
+                            res => panic!("unexpectd result: {:?}", res),
+                        })
+                        .expect("can spawn thread to run guest");
+
+                    racepoint.wait_and_then(|| {
+                        kill_switch.terminate().expect("can terminate in guest");
+                    });
+
+                    guest.join().expect("guest exits without panic");
+                })
+            }
+        }
+
+        // Test a termination that completes right before `exit_guest_region` takes ownership of termination.
+        //
+        // This corresponds to a race during the documentation's State B -> State E "due to normal exit"
```

Comment:
😂 Thanks for mirroring your docs.  This helps me immensely with navigating all the test cases.